// rclnodejs/generated/trident_msgs/trident_msgs__srv__GuidanceRequest_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PointWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Point.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GuidanceRequest_ResponseRefStruct = StructType({
  accepted: primitiveTypes.bool,
  reference_position: PointWrapper.refObjectType,
  message: primitiveTypes.string,
});
const GuidanceRequest_ResponseRefArray = ArrayType(GuidanceRequest_ResponseRefStruct);
const GuidanceRequest_ResponseRefStructArray = StructType({
  data: GuidanceRequest_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GuidanceRequest_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GuidanceRequest_ResponseRefStruct(other._refObject.toObject());
      this._acceptedIntialized = true;
      this._wrapperFields.reference_position =  new PointWrapper(other._wrapperFields.reference_position);
      this._messageIntialized = true;
      this._wrapperFields.message =  new StringWrapper(other._wrapperFields.message);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GuidanceRequest_ResponseRefStruct();
    this._acceptedIntialized = false;
    this._wrapperFields.reference_position = new PointWrapper();
    this._messageIntialized = false;
    this._wrapperFields.message = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GuidanceRequest_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GuidanceRequest_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GuidanceRequest_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GuidanceRequest_ResponseRefStructArray
  }
  static get refObjectType() {
    return GuidanceRequest_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._acceptedIntialized) {
        throw new TypeError('Invalid argument: accepted in GuidanceRequest_Response');
      }
      if (!this._messageIntialized) {
        throw new TypeError('Invalid argument: message in GuidanceRequest_Response');
      }
    }
    this._wrapperFields.reference_position.freeze(own, checkConsistency);
    this._refObject.reference_position = this._wrapperFields.reference_position.refObject;
    if (own) {
      this._wrapperFields.message.freeze(own, checkConsistency);
    }
    this._refObject.message = this._wrapperFields.message.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._acceptedIntialized = true;
    this._refObject.accepted = refObject.accepted;
    this._wrapperFields.reference_position.copyRefObject(refObject.reference_position);
    this._messageIntialized = true;
    this._wrapperFields.message.data = refObject.message.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PointWrapper.freeStruct(refObject.reference_position);
    StringWrapper.freeStruct(refObject.message);
  }
  static destoryRawROS(msg) {
    GuidanceRequest_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'trident_msgs', subFolder: 'srv', interfaceName: 'GuidanceRequest_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get accepted() {
    if (!this._acceptedIntialized) {
      return undefined;
    }
    return this._refObject.accepted;
  }
  set accepted(value) {
    this._acceptedIntialized = true;
    this._refObject.accepted = value;
  }
  get reference_position() {
    return this._wrapperFields.reference_position;
  }
  set reference_position(value) {
    if (value instanceof PointWrapper) {
      this._wrapperFields.reference_position.copy(value);
    } else {
      this._wrapperFields.reference_position.copy(new PointWrapper(value));
    }
  }
  get message() {
    if (!this._messageIntialized) {
      return undefined;
    }
    return this._wrapperFields.message.data;
  }
  set message(value) {
    this._messageIntialized = true;
    this._wrapperFields.message.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GuidanceRequest_ResponseRefStruct(refObject.toObject());
    this._acceptedIntialized = true;
    this._wrapperFields.reference_position.copyRefObject(this._refObject.reference_position);
    this._messageIntialized = true;
    this._wrapperFields.message.copyRefObject(this._refObject.message);
  }
  copy(other) {
    this._refObject = new GuidanceRequest_ResponseRefStruct(other._refObject.toObject());
    this._acceptedIntialized = true;
    this._wrapperFields.reference_position.copy(other._wrapperFields.reference_position);
    this._messageIntialized = true;
    this._wrapperFields.message.copy(other._wrapperFields.message);
  }
  static get classType() {
    return GuidanceRequest_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"accepted","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reference_position","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Point","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"message","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"trident_msgs","type":"GuidanceRequest_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GuidanceRequest_Response"};
  }
  hasMember(name) {
    let memberNames = ["accepted","reference_position","message"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GuidanceRequest_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GuidanceRequest_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GuidanceRequest_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GuidanceRequest_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GuidanceRequest_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GuidanceRequest_ResponseRefArray(size);
    this._refObject = new GuidanceRequest_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GuidanceRequest_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GuidanceRequest_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GuidanceRequest_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GuidanceRequest_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GuidanceRequest_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GuidanceRequest_ResponseArrayWrapper;
  }
}
module.exports = GuidanceRequest_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "accepted",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reference_position",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Point",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "message",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "trident_msgs",
    "type": "GuidanceRequest_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GuidanceRequest_Response"
}
*/
