// rclnodejs/generated/trident_msgs/trident_msgs__msg__Mission.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const WaypointWrapper = require('../../generated/trident_msgs/trident_msgs__msg__Waypoint.js');
const MissionRefStruct = StructType({
  waypoints: WaypointWrapper.refObjectArrayType,
});
const MissionRefArray = ArrayType(MissionRefStruct);
const MissionRefStructArray = StructType({
  data: MissionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MissionWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MissionRefStruct(other._refObject.toObject());
      this._wrapperFields.waypoints = WaypointWrapper.createArray();
      this._wrapperFields.waypoints.copy(other._wrapperFields.waypoints);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MissionRefStruct();
    this._wrapperFields.waypoints = WaypointWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MissionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MissionArrayWrapper;
  }
  static get ArrayType() {
    return MissionArrayWrapper;
  }
  static get refObjectArrayType() {
    return MissionRefStructArray
  }
  static get refObjectType() {
    return MissionRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.waypoints.freeze(own, checkConsistency);
    this._refObject.waypoints = this._wrapperFields.waypoints.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.waypoints.copyRefObject(refObject.waypoints);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.waypoints.size != 0) {
      WaypointWrapper.ArrayType.freeArray(refObject.waypoints);
      if (WaypointWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.waypoints, WaypointWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    MissionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'trident_msgs', subFolder: 'msg', interfaceName: 'Mission'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get waypoints() {
    return this._wrapperFields.waypoints;
  }
  set waypoints(value) {
    this._wrapperFields.waypoints.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MissionRefStruct(refObject.toObject());
    this._wrapperFields.waypoints.copyRefObject(this._refObject.waypoints);
  }
  copy(other) {
    this._refObject = new MissionRefStruct(other._refObject.toObject());
    this._wrapperFields.waypoints.copy(other._wrapperFields.waypoints);
  }
  static get classType() {
    return MissionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"waypoints","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"trident_msgs","type":"Waypoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"trident_msgs","type":"Mission","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Mission"};
  }
  hasMember(name) {
    let memberNames = ["waypoints"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MissionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MissionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MissionWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MissionArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MissionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MissionRefArray(size);
    this._refObject = new MissionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MissionWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MissionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MissionArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MissionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MissionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MissionArrayWrapper;
  }
}
module.exports = MissionWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "waypoints",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "trident_msgs",
        "type": "Waypoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "trident_msgs",
    "type": "Mission",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Mission"
}
*/
