// rclnodejs/generated/trident_msgs/trident_msgs__action__HoldPose_Result.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const HoldPose_ResultRefStruct = StructType({
  duration: primitiveTypes.int64,
  message: primitiveTypes.string,
  mean_pose_deviation: PoseWrapper.refObjectType,
});
const HoldPose_ResultRefArray = ArrayType(HoldPose_ResultRefStruct);
const HoldPose_ResultRefStructArray = StructType({
  data: HoldPose_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class HoldPose_ResultWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new HoldPose_ResultRefStruct(other._refObject.toObject());
      this._durationIntialized = true;
      this._messageIntialized = true;
      this._wrapperFields.message =  new StringWrapper(other._wrapperFields.message);
      this._wrapperFields.mean_pose_deviation =  new PoseWrapper(other._wrapperFields.mean_pose_deviation);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new HoldPose_ResultRefStruct();
    this._durationIntialized = false;
    this._messageIntialized = false;
    this._wrapperFields.message = new StringWrapper();
    this._wrapperFields.mean_pose_deviation = new PoseWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new HoldPose_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new HoldPose_ResultArrayWrapper;
  }
  static get ArrayType() {
    return HoldPose_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return HoldPose_ResultRefStructArray
  }
  static get refObjectType() {
    return HoldPose_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._durationIntialized) {
        throw new TypeError('Invalid argument: duration in HoldPose_Result');
      }
      if (!this._messageIntialized) {
        throw new TypeError('Invalid argument: message in HoldPose_Result');
      }
    }
    if (own) {
      this._wrapperFields.message.freeze(own, checkConsistency);
    }
    this._refObject.message = this._wrapperFields.message.refObject;
    this._wrapperFields.mean_pose_deviation.freeze(own, checkConsistency);
    this._refObject.mean_pose_deviation = this._wrapperFields.mean_pose_deviation.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._durationIntialized = true;
    this._refObject.duration = refObject.duration;
    this._messageIntialized = true;
    this._wrapperFields.message.data = refObject.message.data;
    this._wrapperFields.mean_pose_deviation.copyRefObject(refObject.mean_pose_deviation);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.message);
    PoseWrapper.freeStruct(refObject.mean_pose_deviation);
  }
  static destoryRawROS(msg) {
    HoldPose_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'trident_msgs', subFolder: 'action', interfaceName: 'HoldPose_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get duration() {
    if (!this._durationIntialized) {
      return undefined;
    }
    return this._refObject.duration;
  }
  set duration(value) {
    this._durationIntialized = true;
    this._refObject.duration = value;
  }
  get message() {
    if (!this._messageIntialized) {
      return undefined;
    }
    return this._wrapperFields.message.data;
  }
  set message(value) {
    this._messageIntialized = true;
    this._wrapperFields.message.data = value;
  }
  get mean_pose_deviation() {
    return this._wrapperFields.mean_pose_deviation;
  }
  set mean_pose_deviation(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.mean_pose_deviation.copy(value);
    } else {
      this._wrapperFields.mean_pose_deviation.copy(new PoseWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new HoldPose_ResultRefStruct(refObject.toObject());
    this._durationIntialized = true;
    this._messageIntialized = true;
    this._wrapperFields.message.copyRefObject(this._refObject.message);
    this._wrapperFields.mean_pose_deviation.copyRefObject(this._refObject.mean_pose_deviation);
  }
  copy(other) {
    this._refObject = new HoldPose_ResultRefStruct(other._refObject.toObject());
    this._durationIntialized = true;
    this._messageIntialized = true;
    this._wrapperFields.message.copy(other._wrapperFields.message);
    this._wrapperFields.mean_pose_deviation.copy(other._wrapperFields.mean_pose_deviation);
  }
  static get classType() {
    return HoldPose_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"duration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"message","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mean_pose_deviation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"trident_msgs","type":"HoldPose_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"HoldPose_Result"};
  }
  hasMember(name) {
    let memberNames = ["duration","message","mean_pose_deviation"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class HoldPose_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof HoldPose_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new HoldPose_ResultWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HoldPose_ResultArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HoldPose_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new HoldPose_ResultRefArray(size);
    this._refObject = new HoldPose_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new HoldPose_ResultWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof HoldPose_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "HoldPose_ResultArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      HoldPose_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return HoldPose_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return HoldPose_ResultArrayWrapper;
  }
}
module.exports = HoldPose_ResultWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "duration",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "message",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mean_pose_deviation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "trident_msgs",
    "type": "HoldPose_Result",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "HoldPose_Result"
}
*/
