// rclnodejs/generated/trident_msgs/trident_msgs__action__GotoWaypoint_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const WaypointWrapper = require('../../generated/trident_msgs/trident_msgs__msg__Waypoint.js');
const GotoWaypoint_GoalRefStruct = StructType({
  waypoint: WaypointWrapper.refObjectType,
});
const GotoWaypoint_GoalRefArray = ArrayType(GotoWaypoint_GoalRefStruct);
const GotoWaypoint_GoalRefStructArray = StructType({
  data: GotoWaypoint_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GotoWaypoint_GoalWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GotoWaypoint_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.waypoint =  new WaypointWrapper(other._wrapperFields.waypoint);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GotoWaypoint_GoalRefStruct();
    this._wrapperFields.waypoint = new WaypointWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GotoWaypoint_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GotoWaypoint_GoalArrayWrapper;
  }
  static get ArrayType() {
    return GotoWaypoint_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return GotoWaypoint_GoalRefStructArray
  }
  static get refObjectType() {
    return GotoWaypoint_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.waypoint.freeze(own, checkConsistency);
    this._refObject.waypoint = this._wrapperFields.waypoint.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.waypoint.copyRefObject(refObject.waypoint);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    WaypointWrapper.freeStruct(refObject.waypoint);
  }
  static destoryRawROS(msg) {
    GotoWaypoint_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'trident_msgs', subFolder: 'action', interfaceName: 'GotoWaypoint_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get waypoint() {
    return this._wrapperFields.waypoint;
  }
  set waypoint(value) {
    if (value instanceof WaypointWrapper) {
      this._wrapperFields.waypoint.copy(value);
    } else {
      this._wrapperFields.waypoint.copy(new WaypointWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GotoWaypoint_GoalRefStruct(refObject.toObject());
    this._wrapperFields.waypoint.copyRefObject(this._refObject.waypoint);
  }
  copy(other) {
    this._refObject = new GotoWaypoint_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.waypoint.copy(other._wrapperFields.waypoint);
  }
  static get classType() {
    return GotoWaypoint_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"waypoint","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"trident_msgs","type":"Waypoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"trident_msgs","type":"GotoWaypoint_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GotoWaypoint_Goal"};
  }
  hasMember(name) {
    let memberNames = ["waypoint"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GotoWaypoint_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GotoWaypoint_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GotoWaypoint_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GotoWaypoint_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GotoWaypoint_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GotoWaypoint_GoalRefArray(size);
    this._refObject = new GotoWaypoint_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GotoWaypoint_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GotoWaypoint_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GotoWaypoint_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GotoWaypoint_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GotoWaypoint_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GotoWaypoint_GoalArrayWrapper;
  }
}
module.exports = GotoWaypoint_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "waypoint",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "trident_msgs",
        "type": "Waypoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "trident_msgs",
    "type": "GotoWaypoint_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GotoWaypoint_Goal"
}
*/
