// rclnodejs/generated/trident_msgs/trident_msgs__srv__KalmanSensorService_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const KalmanSensorService_ResponseRefStruct = StructType({
  residual: Float32Wrapper.refObjectArrayType,
  gain: Float32Wrapper.refObjectArrayType,
  observationmatrix: Float32Wrapper.refObjectArrayType,
});
const KalmanSensorService_ResponseRefArray = ArrayType(KalmanSensorService_ResponseRefStruct);
const KalmanSensorService_ResponseRefStructArray = StructType({
  data: KalmanSensorService_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class KalmanSensorService_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new KalmanSensorService_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.residual = Float32Wrapper.createArray();
      this._wrapperFields.residual.copy(other._wrapperFields.residual);
      this._wrapperFields.gain = Float32Wrapper.createArray();
      this._wrapperFields.gain.copy(other._wrapperFields.gain);
      this._wrapperFields.observationmatrix = Float32Wrapper.createArray();
      this._wrapperFields.observationmatrix.copy(other._wrapperFields.observationmatrix);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new KalmanSensorService_ResponseRefStruct();
    this._wrapperFields.residual = Float32Wrapper.createArray();
    this._wrapperFields.gain = Float32Wrapper.createArray();
    this._wrapperFields.observationmatrix = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new KalmanSensorService_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new KalmanSensorService_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return KalmanSensorService_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return KalmanSensorService_ResponseRefStructArray
  }
  static get refObjectType() {
    return KalmanSensorService_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.residual.freeze(own, checkConsistency);
    this._refObject.residual = this._wrapperFields.residual.refObject;
    if (own) {
      this._wrapperFields.residual.fill(Float32Array.from([]));
      this._wrapperFields.residual.freeze(own, checkConsistency);
      this._refObject.residual = this._wrapperFields.residual.refObject;
    }
    this._wrapperFields.gain.freeze(own, checkConsistency);
    this._refObject.gain = this._wrapperFields.gain.refObject;
    if (own) {
      this._wrapperFields.gain.fill(Float32Array.from([]));
      this._wrapperFields.gain.freeze(own, checkConsistency);
      this._refObject.gain = this._wrapperFields.gain.refObject;
    }
    this._wrapperFields.observationmatrix.freeze(own, checkConsistency);
    this._refObject.observationmatrix = this._wrapperFields.observationmatrix.refObject;
    if (own) {
      this._wrapperFields.observationmatrix.fill(Float32Array.from([]));
      this._wrapperFields.observationmatrix.freeze(own, checkConsistency);
      this._refObject.observationmatrix = this._wrapperFields.observationmatrix.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.residual.copyRefObject(refObject.residual);
    this._wrapperFields.gain.copyRefObject(refObject.gain);
    this._wrapperFields.observationmatrix.copyRefObject(refObject.observationmatrix);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.residual.size != 0) {
      Float32Wrapper.ArrayType.freeArray(refObject.residual);
      if (Float32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.residual, Float32Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.gain.size != 0) {
      Float32Wrapper.ArrayType.freeArray(refObject.gain);
      if (Float32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.gain, Float32Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.observationmatrix.size != 0) {
      Float32Wrapper.ArrayType.freeArray(refObject.observationmatrix);
      if (Float32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.observationmatrix, Float32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    KalmanSensorService_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'trident_msgs', subFolder: 'srv', interfaceName: 'KalmanSensorService_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get residual() {
    return this._wrapperFields['residual'].data;
  }
  set residual(value) {
    this._wrapperFields['residual'].fill(value);
  }
  get gain() {
    return this._wrapperFields['gain'].data;
  }
  set gain(value) {
    this._wrapperFields['gain'].fill(value);
  }
  get observationmatrix() {
    return this._wrapperFields['observationmatrix'].data;
  }
  set observationmatrix(value) {
    this._wrapperFields['observationmatrix'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new KalmanSensorService_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.residual.copyRefObject(this._refObject.residual);
    this._wrapperFields.gain.copyRefObject(this._refObject.gain);
    this._wrapperFields.observationmatrix.copyRefObject(this._refObject.observationmatrix);
  }
  copy(other) {
    this._refObject = new KalmanSensorService_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.residual.copy(other._wrapperFields.residual);
    this._wrapperFields.gain.copy(other._wrapperFields.gain);
    this._wrapperFields.observationmatrix.copy(other._wrapperFields.observationmatrix);
  }
  static get classType() {
    return KalmanSensorService_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"residual","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gain","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"observationmatrix","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"trident_msgs","type":"KalmanSensorService_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"KalmanSensorService_Response"};
  }
  hasMember(name) {
    let memberNames = ["residual","gain","observationmatrix"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class KalmanSensorService_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof KalmanSensorService_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new KalmanSensorService_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to KalmanSensorService_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to KalmanSensorService_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new KalmanSensorService_ResponseRefArray(size);
    this._refObject = new KalmanSensorService_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new KalmanSensorService_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof KalmanSensorService_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "KalmanSensorService_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      KalmanSensorService_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return KalmanSensorService_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return KalmanSensorService_ResponseArrayWrapper;
  }
}
module.exports = KalmanSensorService_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "residual",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gain",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "observationmatrix",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "trident_msgs",
    "type": "KalmanSensorService_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "KalmanSensorService_Response"
}
*/
