// rclnodejs/generated/trident_msgs/trident_msgs__msg__PidParam.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PidParamRefStruct = StructType({
  pid_element: primitiveTypes.string,
  key: primitiveTypes.string,
  value: primitiveTypes.float64,
});
const PidParamRefArray = ArrayType(PidParamRefStruct);
const PidParamRefStructArray = StructType({
  data: PidParamRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PidParamWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PidParamRefStruct(other._refObject.toObject());
      this._pid_elementIntialized = true;
      this._wrapperFields.pid_element =  new StringWrapper(other._wrapperFields.pid_element);
      this._keyIntialized = true;
      this._wrapperFields.key =  new StringWrapper(other._wrapperFields.key);
      this._valueIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PidParamRefStruct();
    this._pid_elementIntialized = false;
    this._wrapperFields.pid_element = new StringWrapper();
    this._keyIntialized = false;
    this._wrapperFields.key = new StringWrapper();
    this._valueIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new PidParamWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PidParamArrayWrapper;
  }
  static get ArrayType() {
    return PidParamArrayWrapper;
  }
  static get refObjectArrayType() {
    return PidParamRefStructArray
  }
  static get refObjectType() {
    return PidParamRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._pid_elementIntialized) {
        throw new TypeError('Invalid argument: pid_element in PidParam');
      }
      if (!this._keyIntialized) {
        throw new TypeError('Invalid argument: key in PidParam');
      }
      if (!this._valueIntialized) {
        throw new TypeError('Invalid argument: value in PidParam');
      }
    }
    if (own) {
      this._wrapperFields.pid_element.freeze(own, checkConsistency);
    }
    this._refObject.pid_element = this._wrapperFields.pid_element.refObject;
    if (own) {
      this._wrapperFields.key.freeze(own, checkConsistency);
    }
    this._refObject.key = this._wrapperFields.key.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._pid_elementIntialized = true;
    this._wrapperFields.pid_element.data = refObject.pid_element.data;
    this._keyIntialized = true;
    this._wrapperFields.key.data = refObject.key.data;
    this._valueIntialized = true;
    this._refObject.value = refObject.value;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.pid_element);
    StringWrapper.freeStruct(refObject.key);
  }
  static destoryRawROS(msg) {
    PidParamWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'trident_msgs', subFolder: 'msg', interfaceName: 'PidParam'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get pid_element() {
    if (!this._pid_elementIntialized) {
      return undefined;
    }
    return this._wrapperFields.pid_element.data;
  }
  set pid_element(value) {
    this._pid_elementIntialized = true;
    this._wrapperFields.pid_element.data = value;
  }
  get key() {
    if (!this._keyIntialized) {
      return undefined;
    }
    return this._wrapperFields.key.data;
  }
  set key(value) {
    this._keyIntialized = true;
    this._wrapperFields.key.data = value;
  }
  get value() {
    if (!this._valueIntialized) {
      return undefined;
    }
    return this._refObject.value;
  }
  set value(value) {
    this._valueIntialized = true;
    this._refObject.value = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PidParamRefStruct(refObject.toObject());
    this._pid_elementIntialized = true;
    this._wrapperFields.pid_element.copyRefObject(this._refObject.pid_element);
    this._keyIntialized = true;
    this._wrapperFields.key.copyRefObject(this._refObject.key);
    this._valueIntialized = true;
  }
  copy(other) {
    this._refObject = new PidParamRefStruct(other._refObject.toObject());
    this._pid_elementIntialized = true;
    this._wrapperFields.pid_element.copy(other._wrapperFields.pid_element);
    this._keyIntialized = true;
    this._wrapperFields.key.copy(other._wrapperFields.key);
    this._valueIntialized = true;
  }
  static get classType() {
    return PidParamWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"pid_element","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"key","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"trident_msgs","type":"PidParam","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PidParam"};
  }
  hasMember(name) {
    let memberNames = ["pid_element","key","value"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PidParamArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PidParamWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PidParamWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PidParamArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PidParamArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PidParamRefArray(size);
    this._refObject = new PidParamRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PidParamWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PidParamArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PidParamArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PidParamWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PidParamWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PidParamArrayWrapper;
  }
}
module.exports = PidParamWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "pid_element",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "key",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "trident_msgs",
    "type": "PidParam",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PidParam"
}
*/
