// rclnodejs/generated/trident_msgs/trident_msgs__action__StartMission_Feedback.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const StartMission_FeedbackRefStruct = StructType({
  status: primitiveTypes.int16,
  message: primitiveTypes.string,
  waypoints_completed: primitiveTypes.int16,
});
const StartMission_FeedbackRefArray = ArrayType(StartMission_FeedbackRefStruct);
const StartMission_FeedbackRefStructArray = StructType({
  data: StartMission_FeedbackRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class StartMission_FeedbackWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new StartMission_FeedbackRefStruct(other._refObject.toObject());
      this._statusIntialized = true;
      this._messageIntialized = true;
      this._wrapperFields.message =  new StringWrapper(other._wrapperFields.message);
      this._waypoints_completedIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new StartMission_FeedbackRefStruct();
    this._statusIntialized = false;
    this._messageIntialized = false;
    this._wrapperFields.message = new StringWrapper();
    this._waypoints_completedIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new StartMission_FeedbackWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new StartMission_FeedbackArrayWrapper;
  }
  static get ArrayType() {
    return StartMission_FeedbackArrayWrapper;
  }
  static get refObjectArrayType() {
    return StartMission_FeedbackRefStructArray
  }
  static get refObjectType() {
    return StartMission_FeedbackRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._statusIntialized) {
        throw new TypeError('Invalid argument: status in StartMission_Feedback');
      }
      if (!this._messageIntialized) {
        throw new TypeError('Invalid argument: message in StartMission_Feedback');
      }
      if (!this._waypoints_completedIntialized) {
        throw new TypeError('Invalid argument: waypoints_completed in StartMission_Feedback');
      }
    }
    if (own) {
      this._wrapperFields.message.freeze(own, checkConsistency);
    }
    this._refObject.message = this._wrapperFields.message.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._statusIntialized = true;
    this._refObject.status = refObject.status;
    this._messageIntialized = true;
    this._wrapperFields.message.data = refObject.message.data;
    this._waypoints_completedIntialized = true;
    this._refObject.waypoints_completed = refObject.waypoints_completed;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.message);
  }
  static destoryRawROS(msg) {
    StartMission_FeedbackWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'trident_msgs', subFolder: 'action', interfaceName: 'StartMission_Feedback'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get status() {
    if (!this._statusIntialized) {
      return undefined;
    }
    return this._refObject.status;
  }
  set status(value) {
    this._statusIntialized = true;
    this._refObject.status = value;
  }
  get message() {
    if (!this._messageIntialized) {
      return undefined;
    }
    return this._wrapperFields.message.data;
  }
  set message(value) {
    this._messageIntialized = true;
    this._wrapperFields.message.data = value;
  }
  get waypoints_completed() {
    if (!this._waypoints_completedIntialized) {
      return undefined;
    }
    return this._refObject.waypoints_completed;
  }
  set waypoints_completed(value) {
    this._waypoints_completedIntialized = true;
    this._refObject.waypoints_completed = value;
  }
  copyRefObject(refObject) {
    this._refObject = new StartMission_FeedbackRefStruct(refObject.toObject());
    this._statusIntialized = true;
    this._messageIntialized = true;
    this._wrapperFields.message.copyRefObject(this._refObject.message);
    this._waypoints_completedIntialized = true;
  }
  copy(other) {
    this._refObject = new StartMission_FeedbackRefStruct(other._refObject.toObject());
    this._statusIntialized = true;
    this._messageIntialized = true;
    this._wrapperFields.message.copy(other._wrapperFields.message);
    this._waypoints_completedIntialized = true;
  }
  static get classType() {
    return StartMission_FeedbackWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"message","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"waypoints_completed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"trident_msgs","type":"StartMission_Feedback","stringUpperBound":null,"isPrimitiveType":false},"msgName":"StartMission_Feedback"};
  }
  hasMember(name) {
    let memberNames = ["status","message","waypoints_completed"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class StartMission_FeedbackArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof StartMission_FeedbackWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new StartMission_FeedbackWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StartMission_FeedbackArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StartMission_FeedbackArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new StartMission_FeedbackRefArray(size);
    this._refObject = new StartMission_FeedbackRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new StartMission_FeedbackWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof StartMission_FeedbackArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "StartMission_FeedbackArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      StartMission_FeedbackWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return StartMission_FeedbackWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return StartMission_FeedbackArrayWrapper;
  }
}
module.exports = StartMission_FeedbackWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "message",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "waypoints_completed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "trident_msgs",
    "type": "StartMission_Feedback",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "StartMission_Feedback"
}
*/
