// rclnodejs/generated/trident_msgs/trident_msgs__srv__GetState_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GetState_ResponseRefStruct = StructType({
  success: primitiveTypes.bool,
  state: primitiveTypes.string,
  int_state: primitiveTypes.int64,
});
const GetState_ResponseRefArray = ArrayType(GetState_ResponseRefStruct);
const GetState_ResponseRefStructArray = StructType({
  data: GetState_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetState_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetState_ResponseRefStruct(other._refObject.toObject());
      this._successIntialized = true;
      this._stateIntialized = true;
      this._wrapperFields.state =  new StringWrapper(other._wrapperFields.state);
      this._int_stateIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetState_ResponseRefStruct();
    this._successIntialized = false;
    this._stateIntialized = false;
    this._wrapperFields.state = new StringWrapper();
    this._refObject.int_state = -1;
    this._int_stateIntialized = true;
  }
  static createFromRefObject(refObject) {
    let self = new GetState_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetState_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetState_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetState_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetState_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._successIntialized) {
        throw new TypeError('Invalid argument: success in GetState_Response');
      }
      if (!this._stateIntialized) {
        throw new TypeError('Invalid argument: state in GetState_Response');
      }
      if (!this._int_stateIntialized) {
        throw new TypeError('Invalid argument: int_state in GetState_Response');
      }
    }
    if (own) {
      this._wrapperFields.state.freeze(own, checkConsistency);
    }
    this._refObject.state = this._wrapperFields.state.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._successIntialized = true;
    this._refObject.success = refObject.success;
    this._stateIntialized = true;
    this._wrapperFields.state.data = refObject.state.data;
    this._int_stateIntialized = true;
    this._refObject.int_state = refObject.int_state;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.state);
  }
  static destoryRawROS(msg) {
    GetState_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'trident_msgs', subFolder: 'srv', interfaceName: 'GetState_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get success() {
    if (!this._successIntialized) {
      return undefined;
    }
    return this._refObject.success;
  }
  set success(value) {
    this._successIntialized = true;
    this._refObject.success = value;
  }
  get state() {
    if (!this._stateIntialized) {
      return undefined;
    }
    return this._wrapperFields.state.data;
  }
  set state(value) {
    this._stateIntialized = true;
    this._wrapperFields.state.data = value;
  }
  get int_state() {
    if (!this._int_stateIntialized) {
      return undefined;
    }
    return this._refObject.int_state;
  }
  set int_state(value) {
    this._int_stateIntialized = true;
    this._refObject.int_state = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetState_ResponseRefStruct(refObject.toObject());
    this._successIntialized = true;
    this._stateIntialized = true;
    this._wrapperFields.state.copyRefObject(this._refObject.state);
    this._int_stateIntialized = true;
  }
  copy(other) {
    this._refObject = new GetState_ResponseRefStruct(other._refObject.toObject());
    this._successIntialized = true;
    this._stateIntialized = true;
    this._wrapperFields.state.copy(other._wrapperFields.state);
    this._int_stateIntialized = true;
  }
  static get classType() {
    return GetState_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"int_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int64","stringUpperBound":null,"isPrimitiveType":true},"default_value":-1}],"baseType":{"pkgName":"trident_msgs","type":"GetState_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetState_Response"};
  }
  hasMember(name) {
    let memberNames = ["success","state","int_state"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetState_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetState_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetState_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetState_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetState_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetState_ResponseRefArray(size);
    this._refObject = new GetState_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetState_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetState_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetState_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetState_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetState_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetState_ResponseArrayWrapper;
  }
}
module.exports = GetState_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "success",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "int_state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": -1
    }
  ],
  "baseType": {
    "pkgName": "trident_msgs",
    "type": "GetState_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetState_Response"
}
*/
