// rclnodejs/generated/trident_msgs/trident_msgs__srv__BasicSensorService_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const BasicSensorService_RequestRefStruct = StructType({
  state: Float32Wrapper.refObjectArrayType,
});
const BasicSensorService_RequestRefArray = ArrayType(BasicSensorService_RequestRefStruct);
const BasicSensorService_RequestRefStructArray = StructType({
  data: BasicSensorService_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class BasicSensorService_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new BasicSensorService_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.state = Float32Wrapper.createArray();
      this._wrapperFields.state.copy(other._wrapperFields.state);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new BasicSensorService_RequestRefStruct();
    this._wrapperFields.state = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new BasicSensorService_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new BasicSensorService_RequestArrayWrapper;
  }
  static get ArrayType() {
    return BasicSensorService_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return BasicSensorService_RequestRefStructArray
  }
  static get refObjectType() {
    return BasicSensorService_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.state.freeze(own, checkConsistency);
    this._refObject.state = this._wrapperFields.state.refObject;
    if (own) {
      this._wrapperFields.state.fill(Float32Array.from([]));
      this._wrapperFields.state.freeze(own, checkConsistency);
      this._refObject.state = this._wrapperFields.state.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.state.copyRefObject(refObject.state);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.state.size != 0) {
      Float32Wrapper.ArrayType.freeArray(refObject.state);
      if (Float32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.state, Float32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    BasicSensorService_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'trident_msgs', subFolder: 'srv', interfaceName: 'BasicSensorService_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get state() {
    return this._wrapperFields['state'].data;
  }
  set state(value) {
    this._wrapperFields['state'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new BasicSensorService_RequestRefStruct(refObject.toObject());
    this._wrapperFields.state.copyRefObject(this._refObject.state);
  }
  copy(other) {
    this._refObject = new BasicSensorService_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.state.copy(other._wrapperFields.state);
  }
  static get classType() {
    return BasicSensorService_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"state","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"trident_msgs","type":"BasicSensorService_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"BasicSensorService_Request"};
  }
  hasMember(name) {
    let memberNames = ["state"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class BasicSensorService_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof BasicSensorService_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new BasicSensorService_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BasicSensorService_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BasicSensorService_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new BasicSensorService_RequestRefArray(size);
    this._refObject = new BasicSensorService_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new BasicSensorService_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof BasicSensorService_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "BasicSensorService_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      BasicSensorService_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return BasicSensorService_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return BasicSensorService_RequestArrayWrapper;
  }
}
module.exports = BasicSensorService_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "state",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "trident_msgs",
    "type": "BasicSensorService_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "BasicSensorService_Request"
}
*/
