// rclnodejs/generated/trident_msgs/trident_msgs__srv__GuidanceRequest_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GuidanceRequest_RequestRefStruct = StructType({
  duration: primitiveTypes.int64,
});
const GuidanceRequest_RequestRefArray = ArrayType(GuidanceRequest_RequestRefStruct);
const GuidanceRequest_RequestRefStructArray = StructType({
  data: GuidanceRequest_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GuidanceRequest_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GuidanceRequest_RequestRefStruct(other._refObject.toObject());
      this._durationIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GuidanceRequest_RequestRefStruct();
    this._durationIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GuidanceRequest_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GuidanceRequest_RequestArrayWrapper;
  }
  static get ArrayType() {
    return GuidanceRequest_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GuidanceRequest_RequestRefStructArray
  }
  static get refObjectType() {
    return GuidanceRequest_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._durationIntialized) {
        throw new TypeError('Invalid argument: duration in GuidanceRequest_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._durationIntialized = true;
    this._refObject.duration = refObject.duration;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    GuidanceRequest_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'trident_msgs', subFolder: 'srv', interfaceName: 'GuidanceRequest_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get duration() {
    if (!this._durationIntialized) {
      return undefined;
    }
    return this._refObject.duration;
  }
  set duration(value) {
    this._durationIntialized = true;
    this._refObject.duration = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GuidanceRequest_RequestRefStruct(refObject.toObject());
    this._durationIntialized = true;
  }
  copy(other) {
    this._refObject = new GuidanceRequest_RequestRefStruct(other._refObject.toObject());
    this._durationIntialized = true;
  }
  static get classType() {
    return GuidanceRequest_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"duration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"trident_msgs","type":"GuidanceRequest_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GuidanceRequest_Request"};
  }
  hasMember(name) {
    let memberNames = ["duration"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GuidanceRequest_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GuidanceRequest_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GuidanceRequest_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GuidanceRequest_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GuidanceRequest_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GuidanceRequest_RequestRefArray(size);
    this._refObject = new GuidanceRequest_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GuidanceRequest_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GuidanceRequest_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GuidanceRequest_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GuidanceRequest_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GuidanceRequest_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GuidanceRequest_RequestArrayWrapper;
  }
}
module.exports = GuidanceRequest_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "duration",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "trident_msgs",
    "type": "GuidanceRequest_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GuidanceRequest_Request"
}
*/
