// rclnodejs/generated/trident_msgs/trident_msgs__action__HoldPose.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class trident_msgs__action__HoldPose {
  static get Goal() {
    return require('./trident_msgs__action__HoldPose_Goal.js');
  }
  static get Result() {
    return require('./trident_msgs__action__HoldPose_Result.js');
  }
  static get Feedback() {
    return require('./trident_msgs__action__HoldPose_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./trident_msgs__action__HoldPose_SendGoal.js');
      },
      get GetResultService() {
        return require('./trident_msgs__action__HoldPose_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./trident_msgs__action__HoldPose_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'trident_msgs', subFolder: 'action', interfaceName: 'HoldPose'};
  }
}
module.exports = trident_msgs__action__HoldPose;
