// rclnodejs/generated/trident_msgs/trident_msgs__msg__WaypointAction.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const WaypointActionRefStruct = StructType({
  action_type: primitiveTypes.int16,
  action_param: primitiveTypes.int16,
});
const WaypointActionRefArray = ArrayType(WaypointActionRefStruct);
const WaypointActionRefStructArray = StructType({
  data: WaypointActionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class WaypointActionWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new WaypointActionRefStruct(other._refObject.toObject());
      this._action_typeIntialized = true;
      this._action_paramIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new WaypointActionRefStruct();
    this._refObject.action_type = 0;
    this._action_typeIntialized = true;
    this._refObject.action_param = 0;
    this._action_paramIntialized = true;
  }
  static createFromRefObject(refObject) {
    let self = new WaypointActionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new WaypointActionArrayWrapper;
  }
  static get ArrayType() {
    return WaypointActionArrayWrapper;
  }
  static get refObjectArrayType() {
    return WaypointActionRefStructArray
  }
  static get refObjectType() {
    return WaypointActionRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._action_typeIntialized) {
        throw new TypeError('Invalid argument: action_type in WaypointAction');
      }
      if (!this._action_paramIntialized) {
        throw new TypeError('Invalid argument: action_param in WaypointAction');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._action_typeIntialized = true;
    this._refObject.action_type = refObject.action_type;
    this._action_paramIntialized = true;
    this._refObject.action_param = refObject.action_param;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    WaypointActionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'trident_msgs', subFolder: 'msg', interfaceName: 'WaypointAction'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get action_type() {
    if (!this._action_typeIntialized) {
      return undefined;
    }
    return this._refObject.action_type;
  }
  set action_type(value) {
    this._action_typeIntialized = true;
    this._refObject.action_type = value;
  }
  get action_param() {
    if (!this._action_paramIntialized) {
      return undefined;
    }
    return this._refObject.action_param;
  }
  set action_param(value) {
    this._action_paramIntialized = true;
    this._refObject.action_param = value;
  }
  copyRefObject(refObject) {
    this._refObject = new WaypointActionRefStruct(refObject.toObject());
    this._action_typeIntialized = true;
    this._action_paramIntialized = true;
  }
  copy(other) {
    this._refObject = new WaypointActionRefStruct(other._refObject.toObject());
    this._action_typeIntialized = true;
    this._action_paramIntialized = true;
  }
  static get classType() {
    return WaypointActionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"action_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":0},{"name":"action_param","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":0}],"baseType":{"pkgName":"trident_msgs","type":"WaypointAction","stringUpperBound":null,"isPrimitiveType":false},"msgName":"WaypointAction"};
  }
  hasMember(name) {
    let memberNames = ["action_type","action_param"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class WaypointActionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof WaypointActionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new WaypointActionWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WaypointActionArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WaypointActionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new WaypointActionRefArray(size);
    this._refObject = new WaypointActionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new WaypointActionWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof WaypointActionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "WaypointActionArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      WaypointActionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return WaypointActionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return WaypointActionArrayWrapper;
  }
}
module.exports = WaypointActionWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "action_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": 0
    },
    {
      "name": "action_param",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": 0
    }
  ],
  "baseType": {
    "pkgName": "trident_msgs",
    "type": "WaypointAction",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "WaypointAction"
}
*/
