// rclnodejs/generated/trident_msgs/trident_msgs__srv__ToggleControl_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ToggleControl_RequestRefStruct = StructType({
  target: primitiveTypes.string,
  mode: primitiveTypes.string,
});
const ToggleControl_RequestRefArray = ArrayType(ToggleControl_RequestRefStruct);
const ToggleControl_RequestRefStructArray = StructType({
  data: ToggleControl_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ToggleControl_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ToggleControl_RequestRefStruct(other._refObject.toObject());
      this._targetIntialized = true;
      this._wrapperFields.target =  new StringWrapper(other._wrapperFields.target);
      this._modeIntialized = true;
      this._wrapperFields.mode =  new StringWrapper(other._wrapperFields.mode);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ToggleControl_RequestRefStruct();
    this._targetIntialized = false;
    this._wrapperFields.target = new StringWrapper();
    this._modeIntialized = false;
    this._wrapperFields.mode = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ToggleControl_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ToggleControl_RequestArrayWrapper;
  }
  static get ArrayType() {
    return ToggleControl_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ToggleControl_RequestRefStructArray
  }
  static get refObjectType() {
    return ToggleControl_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._targetIntialized) {
        throw new TypeError('Invalid argument: target in ToggleControl_Request');
      }
      if (!this._modeIntialized) {
        throw new TypeError('Invalid argument: mode in ToggleControl_Request');
      }
    }
    if (own) {
      this._wrapperFields.target.freeze(own, checkConsistency);
    }
    this._refObject.target = this._wrapperFields.target.refObject;
    if (own) {
      this._wrapperFields.mode.freeze(own, checkConsistency);
    }
    this._refObject.mode = this._wrapperFields.mode.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._targetIntialized = true;
    this._wrapperFields.target.data = refObject.target.data;
    this._modeIntialized = true;
    this._wrapperFields.mode.data = refObject.mode.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.target);
    StringWrapper.freeStruct(refObject.mode);
  }
  static destoryRawROS(msg) {
    ToggleControl_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'trident_msgs', subFolder: 'srv', interfaceName: 'ToggleControl_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get target() {
    if (!this._targetIntialized) {
      return undefined;
    }
    return this._wrapperFields.target.data;
  }
  set target(value) {
    this._targetIntialized = true;
    this._wrapperFields.target.data = value;
  }
  get mode() {
    if (!this._modeIntialized) {
      return undefined;
    }
    return this._wrapperFields.mode.data;
  }
  set mode(value) {
    this._modeIntialized = true;
    this._wrapperFields.mode.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ToggleControl_RequestRefStruct(refObject.toObject());
    this._targetIntialized = true;
    this._wrapperFields.target.copyRefObject(this._refObject.target);
    this._modeIntialized = true;
    this._wrapperFields.mode.copyRefObject(this._refObject.mode);
  }
  copy(other) {
    this._refObject = new ToggleControl_RequestRefStruct(other._refObject.toObject());
    this._targetIntialized = true;
    this._wrapperFields.target.copy(other._wrapperFields.target);
    this._modeIntialized = true;
    this._wrapperFields.mode.copy(other._wrapperFields.mode);
  }
  static get classType() {
    return ToggleControl_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"target","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"trident_msgs","type":"ToggleControl_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ToggleControl_Request"};
  }
  hasMember(name) {
    let memberNames = ["target","mode"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ToggleControl_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ToggleControl_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ToggleControl_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ToggleControl_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ToggleControl_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ToggleControl_RequestRefArray(size);
    this._refObject = new ToggleControl_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ToggleControl_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ToggleControl_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ToggleControl_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ToggleControl_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ToggleControl_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ToggleControl_RequestArrayWrapper;
  }
}
module.exports = ToggleControl_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "target",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "trident_msgs",
    "type": "ToggleControl_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ToggleControl_Request"
}
*/
