// rclnodejs/generated/trident_msgs/trident_msgs__msg__MotorOutputs.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MotorOutputWrapper = require('../../generated/trident_msgs/trident_msgs__msg__MotorOutput.js');
const MotorOutputsRefStruct = StructType({
  motor_outputs: MotorOutputWrapper.refObjectArrayType,
});
const MotorOutputsRefArray = ArrayType(MotorOutputsRefStruct);
const MotorOutputsRefStructArray = StructType({
  data: MotorOutputsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MotorOutputsWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MotorOutputsRefStruct(other._refObject.toObject());
      this._wrapperFields.motor_outputs = MotorOutputWrapper.createArray();
      this._wrapperFields.motor_outputs.copy(other._wrapperFields.motor_outputs);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MotorOutputsRefStruct();
    this._wrapperFields.motor_outputs = MotorOutputWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MotorOutputsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MotorOutputsArrayWrapper;
  }
  static get ArrayType() {
    return MotorOutputsArrayWrapper;
  }
  static get refObjectArrayType() {
    return MotorOutputsRefStructArray
  }
  static get refObjectType() {
    return MotorOutputsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.motor_outputs.freeze(own, checkConsistency);
    this._refObject.motor_outputs = this._wrapperFields.motor_outputs.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.motor_outputs.copyRefObject(refObject.motor_outputs);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.motor_outputs.size != 0) {
      MotorOutputWrapper.ArrayType.freeArray(refObject.motor_outputs);
      if (MotorOutputWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.motor_outputs, MotorOutputWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    MotorOutputsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'trident_msgs', subFolder: 'msg', interfaceName: 'MotorOutputs'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get motor_outputs() {
    return this._wrapperFields.motor_outputs;
  }
  set motor_outputs(value) {
    this._wrapperFields.motor_outputs.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MotorOutputsRefStruct(refObject.toObject());
    this._wrapperFields.motor_outputs.copyRefObject(this._refObject.motor_outputs);
  }
  copy(other) {
    this._refObject = new MotorOutputsRefStruct(other._refObject.toObject());
    this._wrapperFields.motor_outputs.copy(other._wrapperFields.motor_outputs);
  }
  static get classType() {
    return MotorOutputsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"motor_outputs","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"trident_msgs","type":"MotorOutput","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"trident_msgs","type":"MotorOutputs","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MotorOutputs"};
  }
  hasMember(name) {
    let memberNames = ["motor_outputs"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MotorOutputsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MotorOutputsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MotorOutputsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotorOutputsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotorOutputsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MotorOutputsRefArray(size);
    this._refObject = new MotorOutputsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MotorOutputsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MotorOutputsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MotorOutputsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MotorOutputsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MotorOutputsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MotorOutputsArrayWrapper;
  }
}
module.exports = MotorOutputsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "motor_outputs",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "trident_msgs",
        "type": "MotorOutput",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "trident_msgs",
    "type": "MotorOutputs",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MotorOutputs"
}
*/
