// rclnodejs/generated/move_base_msgs/move_base_msgs__action__MoveBase.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class move_base_msgs__action__MoveBase {
  static get Goal() {
    return require('./move_base_msgs__action__MoveBase_Goal.js');
  }
  static get Result() {
    return require('./move_base_msgs__action__MoveBase_Result.js');
  }
  static get Feedback() {
    return require('./move_base_msgs__action__MoveBase_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./move_base_msgs__action__MoveBase_SendGoal.js');
      },
      get GetResultService() {
        return require('./move_base_msgs__action__MoveBase_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./move_base_msgs__action__MoveBase_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'move_base_msgs', subFolder: 'action', interfaceName: 'MoveBase'};
  }
}
module.exports = move_base_msgs__action__MoveBase;
