// rclnodejs/generated/move_base_msgs/move_base_msgs__action__MoveBase_Feedback.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseStamped.js');
const MoveBase_FeedbackRefStruct = StructType({
  base_position: PoseStampedWrapper.refObjectType,
});
const MoveBase_FeedbackRefArray = ArrayType(MoveBase_FeedbackRefStruct);
const MoveBase_FeedbackRefStructArray = StructType({
  data: MoveBase_FeedbackRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MoveBase_FeedbackWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MoveBase_FeedbackRefStruct(other._refObject.toObject());
      this._wrapperFields.base_position =  new PoseStampedWrapper(other._wrapperFields.base_position);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MoveBase_FeedbackRefStruct();
    this._wrapperFields.base_position = new PoseStampedWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new MoveBase_FeedbackWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MoveBase_FeedbackArrayWrapper;
  }
  static get ArrayType() {
    return MoveBase_FeedbackArrayWrapper;
  }
  static get refObjectArrayType() {
    return MoveBase_FeedbackRefStructArray
  }
  static get refObjectType() {
    return MoveBase_FeedbackRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.base_position.freeze(own, checkConsistency);
    this._refObject.base_position = this._wrapperFields.base_position.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.base_position.copyRefObject(refObject.base_position);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PoseStampedWrapper.freeStruct(refObject.base_position);
  }
  static destoryRawROS(msg) {
    MoveBase_FeedbackWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'move_base_msgs', subFolder: 'action', interfaceName: 'MoveBase_Feedback'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get base_position() {
    return this._wrapperFields.base_position;
  }
  set base_position(value) {
    if (value instanceof PoseStampedWrapper) {
      this._wrapperFields.base_position.copy(value);
    } else {
      this._wrapperFields.base_position.copy(new PoseStampedWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new MoveBase_FeedbackRefStruct(refObject.toObject());
    this._wrapperFields.base_position.copyRefObject(this._refObject.base_position);
  }
  copy(other) {
    this._refObject = new MoveBase_FeedbackRefStruct(other._refObject.toObject());
    this._wrapperFields.base_position.copy(other._wrapperFields.base_position);
  }
  static get classType() {
    return MoveBase_FeedbackWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"base_position","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"move_base_msgs","type":"MoveBase_Feedback","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MoveBase_Feedback"};
  }
  hasMember(name) {
    let memberNames = ["base_position"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MoveBase_FeedbackArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MoveBase_FeedbackWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MoveBase_FeedbackWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MoveBase_FeedbackArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MoveBase_FeedbackArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MoveBase_FeedbackRefArray(size);
    this._refObject = new MoveBase_FeedbackRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MoveBase_FeedbackWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MoveBase_FeedbackArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MoveBase_FeedbackArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MoveBase_FeedbackWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MoveBase_FeedbackWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MoveBase_FeedbackArrayWrapper;
  }
}
module.exports = MoveBase_FeedbackWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "base_position",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "PoseStamped",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "move_base_msgs",
    "type": "MoveBase_Feedback",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MoveBase_Feedback"
}
*/
