// rclnodejs/generated/move_base_msgs/move_base_msgs__action__MoveBase_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseStamped.js');
const MoveBase_GoalRefStruct = StructType({
  target_pose: PoseStampedWrapper.refObjectType,
});
const MoveBase_GoalRefArray = ArrayType(MoveBase_GoalRefStruct);
const MoveBase_GoalRefStructArray = StructType({
  data: MoveBase_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MoveBase_GoalWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MoveBase_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.target_pose =  new PoseStampedWrapper(other._wrapperFields.target_pose);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MoveBase_GoalRefStruct();
    this._wrapperFields.target_pose = new PoseStampedWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new MoveBase_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MoveBase_GoalArrayWrapper;
  }
  static get ArrayType() {
    return MoveBase_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return MoveBase_GoalRefStructArray
  }
  static get refObjectType() {
    return MoveBase_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.target_pose.freeze(own, checkConsistency);
    this._refObject.target_pose = this._wrapperFields.target_pose.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.target_pose.copyRefObject(refObject.target_pose);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PoseStampedWrapper.freeStruct(refObject.target_pose);
  }
  static destoryRawROS(msg) {
    MoveBase_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'move_base_msgs', subFolder: 'action', interfaceName: 'MoveBase_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get target_pose() {
    return this._wrapperFields.target_pose;
  }
  set target_pose(value) {
    if (value instanceof PoseStampedWrapper) {
      this._wrapperFields.target_pose.copy(value);
    } else {
      this._wrapperFields.target_pose.copy(new PoseStampedWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new MoveBase_GoalRefStruct(refObject.toObject());
    this._wrapperFields.target_pose.copyRefObject(this._refObject.target_pose);
  }
  copy(other) {
    this._refObject = new MoveBase_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.target_pose.copy(other._wrapperFields.target_pose);
  }
  static get classType() {
    return MoveBase_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"target_pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"move_base_msgs","type":"MoveBase_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MoveBase_Goal"};
  }
  hasMember(name) {
    let memberNames = ["target_pose"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MoveBase_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MoveBase_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MoveBase_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MoveBase_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MoveBase_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MoveBase_GoalRefArray(size);
    this._refObject = new MoveBase_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MoveBase_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MoveBase_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MoveBase_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MoveBase_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MoveBase_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MoveBase_GoalArrayWrapper;
  }
}
module.exports = MoveBase_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "target_pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "PoseStamped",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "move_base_msgs",
    "type": "MoveBase_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MoveBase_Goal"
}
*/
